src/LanguageServer/Protocol/Handler/PullHandlers/VersionedPullCache.CacheItem.cs

```diff
@@ -0,0 +1,139 @@
+            {
+                TCheapVersion cheapVersion;
+                TExpensiveVersion expensiveVersion;
+
+                // Check if the version we have in the cache matches the request version.  If so we can re-use the resultId.
+                if (isFullyLoaded &&
+                    _lastResult is not null &&
+                    _lastResult.Value.resultId == previousPullResult?.PreviousResultId)
+                {
+                    cheapVersion = await cache.ComputeCheapVersionAsync(state, cancellationToken).ConfigureAwait(false);
+                    if (cheapVersion != null && cheapVersion.Equals(_lastResult.Value.cheapVersion))
+                    {
+                        // The client's resultId matches our cached resultId and the cheap version is an
+                        // exact match for our current cheap version. We return early here to avoid calculating
+                        // expensive versions as we know nothing is changed.
+                        return null;
+                    }
+
+                    // The current cheap version does not match the last reported.  This may be because we've forked
+                    // or reloaded a project, so fall back to calculating the full expensive version to determine if
+                    // anything is actually changed.
+                    expensiveVersion = await cache.ComputeExpensiveVersionAsync(state, cancellationToken).ConfigureAwait(false);
+                    if (expensiveVersion != null && expensiveVersion.Equals(_lastResult.Value.expensiveVersion))
+                    {
+                        return null;
+                    }
+                }
+                else
+                {
+                    // The versions we have in our cache (if any) do not match the ones provided by the client (if any).
+                    // We need to calculate new results.
+                    cheapVersion = await cache.ComputeCheapVersionAsync(state, cancellationToken).ConfigureAwait(false);
+                    expensiveVersion = await cache.ComputeExpensiveVersionAsync(state, cancellationToken).ConfigureAwait(false);
+                }
+
+                // Compute the new result for the request.
+                var data = await cache.ComputeDataAsync(state, cancellationToken).ConfigureAwait(false);
+                var dataHash = GetComputedDataHash(data);
+
+                string newResultId;
+                if (_lastResult is not null && _lastResult?.resultId == previousPullResult?.PreviousResultId && _lastResult?.hashedData == dataHash)
+                {
+                    // The new data we've computed is exactly the same as the data we computed last time even though the versions have changed.
+                    // Instead of reserializing everything, we can return the same result id back to the client.
+
+                    // Ensure we store the updated versions we calculated with the old resultId.
```

Comment:
could you mention if this update is actually necessary or not?   wasn't clear to me if that was the case.